pkg/storer/reserve.go

```diff
@@ -172,69 +172,71 @@ func (db *DB) evictionWorker(ctx context.Context) {
 	}
 
 	cleanupExpired := func() {
-		db.metrics.ExpiryTriggersCount.Inc()
 		if !expirySem.TryAcquire(1) {
 			// if there is already a goroutine taking care of expirations dont wait
 			// for it to finish. Cleanup is called from all cases, so we dont
 			// need to trigger everytime.
 			return
 		}
-		defer expirySem.Release(1)
 
-		batchesToEvict, err := db.getExpiredBatches()
-		if err != nil {
-			db.logger.Error(err, "get expired batches")
-			return
-		}
-
-		if len(batchesToEvict) == 0 {
-			return
-		}
+		go func() {
+			defer expirySem.Release(1)
 
-		// we ensure unreserve is not running and if it is we cancel it and wait
-		// for it to finish, this is to prevent unreserve and expirations running
-		// at the same time. The expiration will free up space so the unreserve
-		// target might change. This is to prevent unreserve from running with
-		// the old target.
-		if !unreserveSem.TryAcquire(1) {
-			cancelUnreserve()
-			err := unreserveSem.Acquire(ctx, 1)
+			batchesToEvict, err := db.getExpiredBatches()
 			if err != nil {
-				db.logger.Error(err, "acquire unreserve semaphore")
+				db.logger.Error(err, "get expired batches")
 				return
 			}
-			// trigger it again at the end if required
-			defer db.events.Trigger(reserveOverCapacity)
-		}
-		defer unreserveSem.Release(1)
-
-		defer db.events.Trigger(reserveUnreserved)
 
-		db.metrics.ExpiryRunsCount.Inc()
-
-		for _, batchID := range batchesToEvict {
-			b := batchID
-			if err := expiryWorkers.Acquire(ctx, 1); err != nil {
-				db.logger.Error(err, "acquire expiry worker semaphore")
+			if len(batchesToEvict) == 0 {
 				return
 			}
-			go func() {
-				defer expiryWorkers.Release(1)
 
-				err := db.removeExpiredBatch(ctx, b)
+			// we ensure unreserve is not running and if it is we cancel it and wait
+			// for it to finish, this is to prevent unreserve and expirations running
+			// at the same time. The expiration will free up space so the unreserve
+			// target might change. This is to prevent unreserve from running with
+			// the old target.
+			if !unreserveSem.TryAcquire(1) {
+				cancelUnreserve()
```

Comment:
instead of cancelling, would it make sense to wait for the unreserve to finish first with `unreserveSem.Acquire` ?

Reply:
So it could happen that the unreserve target is substantial. Which is why this cancellation was added to mainly prevent removing valid chunks till we need to.

Reply:
as long as canceling is safe, im ok with this.
another small issue maybe this cancel func needs to be initially assigned a noOp func to prevent a panic

Reply:
This was actually tested. We even found that there was actually an issue with the counts. This is fixed in the next PR where we keep track of the counts per batch. That [fix](https://github.com/ethersphere/bee/pull/4235/files#diff-e37eed2ab5dfe1533be0257473f5b76738361f02d2be71da1b7bf4d5216a00c7R321) was tested with @ldeffenb s help, where the counts were verified after restarts.